client/build_test.go

```diff
@@ -718,6 +722,212 @@ func testClientGatewayContainerMounts(t *testing.T, sb integration.Sandbox) {
 	checkAllReleasable(t, c, sb, true)
 }
 
+// testClientGatewayContainerPID1Tty is testing that we can get a tty via
+// a container pid1, executor.Run
+func testClientGatewayContainerPID1Tty(t *testing.T, sb integration.Sandbox) {
+	requiresLinux(t)
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	product := "buildkit_test"
+
+	inputR, inputW := io.Pipe()
+	output := bytes.NewBuffer(nil)
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		st := llb.Image("busybox:latest")
+
+		def, err := st.Marshal(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal state")
+		}
+
+		r, err := c.Solve(ctx, client.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to solve")
+		}
+
+		ctr, err := c.NewContainer(ctx, client.NewContainerRequest{
+			Mounts: []client.Mount{{
+				Dest:      "/",
+				MountType: pb.MountType_BIND,
+				Ref:       r.Ref,
+			}},
+		})
+		require.NoError(t, err)
+		defer ctr.Release(ctx)
+
+		pid1, err := ctr.Start(ctx, client.StartRequest{
+			Args:   []string{"sh"},
+			Tty:    true,
+			Stdin:  inputR,
+			Stdout: &nopCloser{output},
+			Stderr: &nopCloser{output},
+			Env:    []string{"PS1=% "},
```

Comment:
Just curious, what is this for?

Reply:
Just sets the prompt to constant `% ` to prevent the output from changing if the the defaults change in the future.  The default prompt currently echos the directory first in `sh` on busybox, and echos ansi color codes on alpine, so I was just trying to make it consistent for this test.  It was more useful originally when I was trying to parse the output per prompt, but I gave up on that without a vt10x terminal to process the terminal escape sequences, so it is less useful now.

Reply:
did you try getting rid of the prompt per https://github.com/moby/buildkit/pull/1731#discussion_r506068654 to remove the sleeps per line?

Reply:
I am working on adding a simple output scanner looking for the prompt instead of the sleeps, it seems to work fine.   The expected output is a lot more confusing with the `stty -echo` so I think it works fine to go without that.  We will still have the sleeps to allow the resize signals to go through, 500ms is pessimistic I think, but these tests systems can be slow.

I am now trying to work though an issue with the test hangs (`ctrProc.Wait` hangs) when the expected prompt does not show up in time (ie context cancelled).  It appears to be waiting for the container process Done signal even after `ctr.Release()` is called.  I send another update when I get to the bottom of this.

Reply:
> The expected output is a lot more confusing with the stty -echo so I think it works fine to go without that.

Reply:
> The expected output is a lot more confusing with the stty -echo so I think it works fine to go without that.

What's the new confusing part in the output? I'd expect it to be same, but without repeating all the commands you send. So you can send the commands together and don't need to time between waiting results from the previous command and sending the next one. Still need to wait for `Resize()` calls to trigger but size check can happen multiple times until it eventually gets the correct size or times out.

Reply:
If we leave `% ` the prompt in, with `stty -echo` we will see something like:
```
% stty -echo
% 80 40
% % /tmp
% % foobar
% 100 60
```
vs what we see with the echo on:
```
% ttysize
80 40
% cd /tmp
% pwd
/tmp
% echo foobar > newfile
% cat /tmp/newfile
foobar
% ttysize
100 60
% exit 99
```
We need to leave the prompt in so we can scan for it.  There is only one prompt we really need to wait for, that is the one after the first `ttysize` command, otherwise there is a race between the `ttysize` and the second `ctrProc.Resize(ctx, client.WinSize{Rows: 60, Cols: 100})` so that we know the results of the `ttysize` are from the first resize event.

Reply:
I thought one way to do this is to just:
```
for {
  input.Write([]byte("ttysize\n"))
  time.Sleep(100 * time.Millisecond)
  if strings.Contains(output.String(), expectedSize) {
    break
  } if timeout {
     error
  }
}
```

and not worry about how many times ttysize was called if it eventually switches to correct size before timeout.

Reply:
Sure something like that should work, will update when I figure out what is causing the hanging on timeout errors.

Reply:
I refactored the prompting into a simple helper, it seems cleaner to me.  I also found the hanging issue where if the context was cancelled used for  container `Start` then the event loop was not finishing so stdin continued to be read from.